cpp/serve/radix_tree.cc

```diff
@@ -0,0 +1,438 @@
+    p->parent = page;
+  }
+  page->seq_ids = child->seq_ids;
+  std::vector<int64_t> seq_ids = page->GetLocalSequence();
+  for (int64_t id : seq_ids) seq2page[id] = page;
+  child->seq_ids = nullptr;
+  radix_page_pool->Free(child);
+}
+
+RedixPage* PagedRadixTreeObj::SplitPage(RedixPage* page, size_t offset) {
+  CHECK_LT(offset, page->length);
+  RedixPage* child = radix_page_pool->Allocate();
+  child->parent = page;
+  child->first_child = page->first_child;
+  for (RedixPage* p = page->first_child; p; p = p->next_sibiling) {
+    p->parent = child;
+  }
+  page->first_child = child;
+  for (int i = offset; i < page->length; ++i) {
+    (*child)[i - offset] = (*page)[i];
+  }
+  child->length = page->length - offset;
+  page->length = offset;
+  if (child->Mergeable()) {
+    // The child page may be mergeable
+    MergePage(child);
+  }
+  if (page->parent && page->parent->Mergeable()) {
+    // The parent page may be mergeable
+    page = page->parent;
+    MergePage(page);
+  }
+  return page;
+}
+
+std::tuple<RedixPage*, size_t, size_t> PagedRadixTreeObj::MatchSequence(RedixPage* page,
+                                                                        const int64_t* tokens,
+                                                                        size_t length) {
+  size_t offset = 0;
+  while (offset < length) {
+    if (RedixPage* child = page->FindChild(tokens[offset])) {
+      // If child page starts with offset-th token, common prefix at least ends with child page
+      size_t matched_offset = child->MatchPrefix(tokens + offset, length - offset);
+      offset += matched_offset;
+      if (matched_offset < child->length) {
+        // Common prefix ends within child page
+        return std::make_tuple(child, offset, matched_offset);
+      }
+      page = child;
+    } else {
+      // No child page starts with offset-th token, common prefix ends with current page
+      return std::make_tuple(page, offset, page->length);
+    }
+  }
+  return std::make_tuple(page, length, page->length);
+}
+
+// PagedRadixTree
+
+PagedRadixTree::PagedRadixTree(size_t num_pages, size_t page_size, size_t num_seqs) {
+  ObjectPtr<PagedRadixTreeObj> n = make_object<PagedRadixTreeObj>();
+  n->num_pages = num_pages;
+  n->page_size = page_size;
+  n->num_seqs = num_seqs;
+
+  n->seq_id_node_pool = new SequenceIDNodePool(num_seqs);
+  n->radix_page_pool = new RadixPagePool(page_size, num_pages);
+
+  n->root = reinterpret_cast<RedixPage*>(new int32_t[RedixPage::DATA_OFFSET]);
+  n->root->parent = n->root->first_child = n->root->next_sibiling = nullptr;
+  n->root->offset = n->root->length = n->root->capacity = 0;
+  n->root->seq_ids = nullptr;
+
+  data_ = std::move(n);
+}
+
+TVM_REGISTER_GLOBAL("mlc.serve.PagedRadixTree")
+    .set_body_typed([](uint64_t num_pages, uint64_t page_size, uint64_t num_seqs) {
+      return PagedRadixTree(num_pages, page_size, num_seqs);
+    });
+TVM_REGISTER_GLOBAL("mlc.serve.PagedRadixTree_MatchPrefix")
```

Comment:
```suggestion
TVM_REGISTER_GLOBAL("mlc.serve.PagedRadixTreeMatchPrefix")
```
According to the current naming convention, we do not add a underscore between class name and method name when registering global function. The same below